src/parser/ASTUtils.ts

```diff
@@ -0,0 +1,222 @@
+import { Range, Position } from 'vscode-languageserver';
+import { Statement, Body, AssignmentStatement, Block, ExpressionStatement, CommentStatement, ExitForStatement, ExitWhileStatement, FunctionStatement, IfStatement, IncrementStatement, PrintStatement, GotoStatement, LabelStatement, ReturnStatement, EndStatement, StopStatement, ForStatement, ForEachStatement, WhileStatement, DottedSetStatement, IndexedSetStatement, LibraryStatement, NamespaceStatement, ImportStatement } from './Statement';
+import { Token } from '../lexer/Token';
+import { LiteralExpression, Expression, BinaryExpression, CallExpression, FunctionExpression, NamespacedVariableNameExpression, DottedGetExpression, XmlAttributeGetExpression, IndexedGetExpression, GroupingExpression, EscapedCharCodeLiteral, ArrayLiteralExpression, AALiteralExpression, UnaryExpression, VariableExpression, SourceLiteralExpression, NewExpression, CallfuncExpression, TemplateStringQuasiExpression, TemplateStringExpression, TaggedTemplateStringExpression } from './Expression';
+import { BrsType, BrsString, ValueKind, BrsInvalid, BrsBoolean } from '../brsTypes';
+import { BrsNumber } from '../brsTypes/BrsNumber';
+import { TokenKind } from '../lexer';
+
+// Walking utils
+
+/**
+ * Superficial walk of the statements of a block:
+ * - recurses into if/for/while blocks,
+ * - do not enter sub-functions
+ */
+export function walkStatements(block: Block, visitor: (statement: Statement) => void): void {
+    block?.statements.forEach(s => {
+        visitor(s);
+        if (isIfStatement(s)) {
+            walkStatements(s.thenBranch, visitor);
+            s.elseIfs.forEach(branch => walkStatements(branch.thenBranch, visitor));
+            if (s.elseBranch) {
+                walkStatements(s.elseBranch, visitor);
+            }
+        } else if (hasBody(s)) {
+            walkStatements(s.body, visitor);
+        }
+    });
+}
+
+export function hasBody(statement: Statement): statement is Statement & { body: Block } {
+    return statement && isBlock((statement as any).body);
+}
+
+// Creation utils
+
+export function createRange(pos: Position) {
+    return Range.create(pos.line, pos.character, pos.line, pos.character);
+}
+
+export function createToken(kind: TokenKind, pos: Position, text?: string, literal?: BrsType): Token {
+    return {
+        kind: kind,
+        text: text || kind.toString(),
+        isReserved: !text || text === kind.toString(),
+        range: createRange(pos),
+        literal: literal
+    };
+}
+
+// Statements reflection
```

Comment:
personally; in place of all these functions, I'd rather see an enum on Statement, like statementType, that the statement is responsible for setting. Better encapsulate imo.

Main reason though is working with this will be much nicer with enums than a bunch of if statements.
so if I was doing something I can do
```
switch (statement.type) {
  case StatementType.block:
  case StatementType.exitForStatement:
... etc
```

Reply:
This is the Typescript way to narrow the type of objects - it's very convenient in practice and is the API you get in Typescript compiler plugins itself. Also currently anyway brighterscript statements and expressions don't have a `type` property.

Reply:
that was my point; I thought we should add that property on the expressions and statements. If this is more idiomatic, then it makes sense to follow the idioms: I'm not a ts developer, after all :)